File: java/binary_tree_right_side_view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> res=new ArrayList<Integer>();
        Queue<TreeNode> q=new ArrayDeque<TreeNode>();
        if (root==null) return res;
        q.offer(root);
        int currLevel=1;
        int nextLevel=0;
        
        while (!q.isEmpty()) {
            TreeNode f=q.poll();
            currLevel--;
            if (currLevel==0) res.add(f.val);
            
            if (f.left!=null) {
                q.offer(f.left);
                nextLevel++;
            } 
            
            if (f.right!=null) {
                q.offer(f.right);
                nextLevel++;
            }
            
            if (currLevel==0) {
                currLevel=nextLevel;
                nextLevel=0;
            }
        }
        
        return res;
    }
}

File: python/binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        res=[]
        if not root: return res
        q=[root]
        while q:
            size=len(q)
            res.append(q[0].val)
            
            for i in range(size):
                f=q.pop(0)
                if f.right: q.append(f.right)
                if f.left: q.append(f.left)
        
        return res

